### ISSUE TITLE
Error when using --sudo

### ISSUE THREAD

**Issue by wadey** [ansible/ansible#7147]
##### Issue Type:

Bug Report
##### Ansible Version:

ansible 1.5.5 
##### Environment:

I am running ansible from OS X 10.9.2. I am trying to manage Ubuntu 14.04 running in VirtualBox.
##### Summary:

Running with `--sudo` fails with a weird bash parsing error. I have passwordless sudo enabled in the managed instance.
##### Steps To Reproduce:

run any command with `--sudo` (or with `sudo: yes` in the task)
##### Expected Results:

I expected to see the same results as running without `--sudo`
##### Actual Results:

Here is an example of running without --sudo and running with it.

```
$ ansible --version
ansible 1.5.5

$ ansible all -i hosts -a "echo hello" -vvv
<localhost> ESTABLISH CONNECTION FOR USER: wade
<localhost> REMOTE_MODULE command echo hello
<localhost> EXEC ['ssh', '-C', '-tt', '-q', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=60s', '-o', 'ControlPath=/Users/wade/.ansible/cp/ansible-ssh-%h-%p-%r', '-o', 'Port=2222', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'ConnectTimeout=10', 'localhost', "/bin/sh -c 'mkdir -p /tmp/ansible-tmp-1398318938.25-269243223875077 && chmod a+rx /tmp/ansible-tmp-1398318938.25-269243223875077 && echo /tmp/ansible-tmp-1398318938.25-269243223875077'"]
<localhost> PUT /var/folders/d_/b9p12l191rnf8z0zqnqkd5640000gn/T/tmprVAC2h TO /tmp/ansible-tmp-1398318938.25-269243223875077/command
<localhost> EXEC ['ssh', '-C', '-tt', '-q', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=60s', '-o', 'ControlPath=/Users/wade/.ansible/cp/ansible-ssh-%h-%p-%r', '-o', 'Port=2222', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'ConnectTimeout=10', 'localhost', "/bin/sh -c '/usr/bin/python /tmp/ansible-tmp-1398318938.25-269243223875077/command; rm -rf /tmp/ansible-tmp-1398318938.25-269243223875077/ >/dev/null 2>&1'"]
trusty | success | rc=0 >>
hello

$ ansible all -i hosts -a "echo hello" -vvv --sudo
<localhost> ESTABLISH CONNECTION FOR USER: wade
<localhost> REMOTE_MODULE command echo hello
<localhost> EXEC ['ssh', '-C', '-tt', '-q', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=60s', '-o', 'ControlPath=/Users/wade/.ansible/cp/ansible-ssh-%h-%p-%r', '-o', 'Port=2222', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'ConnectTimeout=10', 'localhost', "/bin/sh -c 'mkdir -p /tmp/ansible-tmp-1398318947.65-140674633396449 && chmod a+rx /tmp/ansible-tmp-1398318947.65-140674633396449 && echo /tmp/ansible-tmp-1398318947.65-140674633396449'"]
<localhost> PUT /var/folders/d_/b9p12l191rnf8z0zqnqkd5640000gn/T/tmpju2Pup TO /tmp/ansible-tmp-1398318947.65-140674633396449/command
<localhost> EXEC ['ssh', '-C', '-tt', '-q', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=60s', '-o', 'ControlPath=/Users/wade/.ansible/cp/ansible-ssh-%h-%p-%r', '-o', 'Port=2222', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'ConnectTimeout=10', 'localhost', '/bin/sh -c \'sudo -k && sudo -H -S -p "[sudo via ansible, key=rwpsduugumxqqyoelowlnqbbfgnjisda] password: " -u root /bin/sh -c \'"\'"\'echo SUDO-SUCCESS-rwpsduugumxqqyoelowlnqbbfgnjisda; /usr/bin/python /tmp/ansible-tmp-1398318947.65-140674633396449/command; rm -rf /tmp/ansible-tmp-1398318947.65-140674633396449/ >/dev/null 2>&1\'"\'"\'\'']
trusty | FAILED >> {
    "failed": true,
    "msg": "bash: -c: line 0: unexpected EOF while looking for matching `\"'\r\nbash: -c: line 1: syntax error: unexpected end of file\r\n",
    "parsed": false
}
```

**Comment by jimi-c** [comment]
I'm not able to reproduce this, though I am connecting from a Fedora system rather than OSX. Can you verify if the issue exists in the current devel branch? I have tested both that and the 1.5.5 version.

**Comment by jimi-c** [comment]
Here is my output:

<pre>
# ansible ubuntu1404 -a "echo hello" -vvv --sudo 
<192.168.122.113> ESTABLISH CONNECTION FOR USER: root
<192.168.122.113> REMOTE_MODULE command echo hello
<192.168.122.113> EXEC ['ssh', '-C', '-q', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=60s', '-o', 'ControlPath=/root/.ansible/cp/%h-%r', '-o', 'StrictHostKeyChecking=no', '-o', 'IdentityFile=/root/.ssh/id_rsa.ansible', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'ConnectTimeout=10', '192.168.122.113', '/bin/sh -c /usr/bin/python']
192.168.122.113 | success | rc=0 >>
hello
</pre>

**Comment by jimi-c** [comment]
I see from the output above that it looks like it's connecting to localhost:2222 for vagrant. Does the same issue occur if you attempt to connect directly to the vagrant instance?

**Comment by wadey** [comment]
@jimi-c in your example, your connecting user is root so it isn't actually using sudo. If you look at the EXEC command it runs you can see that sudo is not in the command at all, where in my example it is running sudo. Can you run a test where you are not connecting as the root user so it uses sudo and see what you get?

I tried multiple versions, from 1.4.5 up to 1.5.5 and all had the same issue. I've tried provisioning two different ubuntu hosts (12.04 and 14.04) and got the same error, so I am pretty sure the issue is on the client side, something to do with OS X or my env.

**Comment by jimi-c** [comment]
I have tested both ways, but since you didn't specify a user in yours I had assumed that was the user you were connecting with. I used a non-root user ('testing') and got the same result, so yes I think it might be something local. 

Can you create a new, clean user account to test from to make sure it's not something to do with your user settings?

**Comment by wadey** [comment]
I will attempt to set up a clean environment to see if I still have an issue.

Can you paste the output you get for the EXEC command when it is using sudo so I can compare?

**Comment by wadey** [comment]
I created a new user account on my local machine, and I am now trying to manage a remote machine instead of a virtualbox instance. The remote machine is a brand new Ubuntu 14.04 instance. I am still getting the same error. Here are the logs with -vvvv showing ssh debug logs:

```
bash-3.2$ ansible all -i hosts -a "echo hello" -vvvv --sudo
<107.170.206.19> ESTABLISH CONNECTION FOR USER: wadetest
<107.170.206.19> REMOTE_MODULE command echo hello
<107.170.206.19> EXEC ['ssh', '-C', '-tt', '-vvv', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=60s', '-o', 'ControlPath=/Users/wadetest/.ansible/cp/ansible-ssh-%h-%p-%r', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'ConnectTimeout=10', '107.170.206.19', "/bin/sh -c 'mkdir -p /tmp/ansible-tmp-1398361722.6-193043791260430 && chmod a+rx /tmp/ansible-tmp-1398361722.6-193043791260430 && echo /tmp/ansible-tmp-1398361722.6-193043791260430'"]
<107.170.206.19> PUT /tmp/tmpXowWmu TO /tmp/ansible-tmp-1398361722.6-193043791260430/command
<107.170.206.19> EXEC ['ssh', '-C', '-tt', '-vvv', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=60s', '-o', 'ControlPath=/Users/wadetest/.ansible/cp/ansible-ssh-%h-%p-%r', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'ConnectTimeout=10', '107.170.206.19', '/bin/sh -c \'sudo -k && sudo -H -S -p "[sudo via ansible, key=koijbvmoumchnynptbthifuouhocdsbr] password: " -u root /bin/sh -c \'"\'"\'echo SUDO-SUCCESS-koijbvmoumchnynptbthifuouhocdsbr; /usr/bin/python /tmp/ansible-tmp-1398361722.6-193043791260430/command; rm -rf /tmp/ansible-tmp-1398361722.6-193043791260430/ >/dev/null 2>&1\'"\'"\'\'']
hub2 | FAILED >> {
    "failed": true,
    "msg": "bash: -c: line 0: unexpected EOF while looking for matching `\"'\r\nbash: -c: line 1: syntax error: unexpected end of file\r\nOpenSSH_6.2p2, OSSLShim 0.9.8r 8 Dec 2011\ndebug1: Reading configuration data /etc/ssh_config\r\ndebug1: /etc/ssh_config line 20: Applying options for *\r\ndebug1: /etc/ssh_config line 102: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 96235\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 1\r\nShared connection to 107.170.206.19 closed.\r\n",
    "parsed": false
}
```

**Comment by jimi-c** [comment]
Here's the output from mine:

<pre>
&lt;192.168.122.113&gt; ESTABLISH CONNECTION FOR USER: testing
&lt;192.168.122.113&gt; REMOTE_MODULE command echo hello
&lt;192.168.122.113&gt; EXEC ['ssh', '-C', '-tt', '-q', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=60s', '-o', 'ControlPath=/root/.ansible/cp/%h-%r', '-o', 'StrictHostKeyChecking=no', '-o', 'IdentityFile=/root/.ssh/id_rsa.ansible', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'User=testing', '-o', 'ConnectTimeout=10', '192.168.122.113', "/bin/sh -c 'mkdir -p $HOME/.ansible/tmp/ansible-tmp-1398362054.75-240576879665494 && chmod a+rx $HOME/.ansible/tmp/ansible-tmp-1398362054.75-240576879665494 && echo $HOME/.ansible/tmp/ansible-tmp-1398362054.75-240576879665494'"]
&lt;192.168.122.113&gt; PUT /tmp/tmp8HCJir TO /home/testing/.ansible/tmp/ansible-tmp-1398362054.75-240576879665494/command
&lt;192.168.122.113&gt; EXEC ['ssh', '-C', '-tt', '-q', '-o', 'ControlMaster=auto', '-o', 'ControlPersist=60s', '-o', 'ControlPath=/root/.ansible/cp/%h-%r', '-o', 'StrictHostKeyChecking=no', '-o', 'IdentityFile=/root/.ssh/id_rsa.ansible', '-o', 'KbdInteractiveAuthentication=no', '-o', 'PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey', '-o', 'PasswordAuthentication=no', '-o', 'User=testing', '-o', 'ConnectTimeout=10', '192.168.122.113', '/bin/sh -c \'sudo -k && sudo -H -S -p "[sudo via ansible, key=mkegeptdeicvxcdqjubxzruebkggtnje] password: " -u root /bin/sh -c \'"\'"\'echo SUDO-SUCCESS-mkegeptdeicvxcdqjubxzruebkggtnje; /usr/bin/python /home/testing/.ansible/tmp/ansible-tmp-1398362054.75-240576879665494/command; rm -rf /home/testing/.ansible/tmp/ansible-tmp-1398362054.75-240576879665494/ &gt;/dev/null 2&gt;&1\'"\'"\'\'']
192.168.122.113 | success | rc=0 &gt;&gt;
hello
</pre>

**Comment by jimi-c** [comment]
If you take the array of args after EXEC and run them through the python interpreter as follows:

<pre>
&gt;&gt;&gt; print " ".join(['ssh', '-C', '-tt', ...]) # everything after EXEC
ssh -C -tt ...
</pre>

Then take that output and run it directly from your console, what do you see? You might need to first run it with ANSIBLE_KEEP_REMOTE_FILES=yes, otherwise the temporary files won't be there anymore.

**Comment by wadey** [comment]
```
bash-3.2$ ssh -C -tt -q -o ControlMaster=auto -o ControlPersist=60s -o ControlPath=/Users/wadetest/.ansible/cp/ansible-ssh-%h-%p-%r -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 107.170.206.19 /bin/sh -c 'sudo -k && sudo -H -S -p "[sudo via ansible, key=vxhanxxypqymccvoeynnnldhqlhnzcyh] password: " -u root /bin/sh -c '"'"'echo SUDO-SUCCESS-vxhanxxypqymccvoeynnnldhqlhnzcyh; /usr/bin/python /tmp/ansible-tmp-1398363267.53-268426956589644/command'"'"''
usage: sudo -h | -K | -k | -V
usage: sudo -v [-AknS] [-g group] [-h host] [-p prompt] [-u user]
usage: sudo -l [-AknS] [-g group] [-h host] [-p prompt] [-U user] [-u user]
            [command]
usage: sudo [-AbEHknPS] [-r role] [-t type] [-C num] [-g group] [-h host] [-p
            prompt] [-u user] [VAR=value] [-i|-s] [<command>]
usage: sudo -e [-AknS] [-r role] [-t type] [-C num] [-g group] [-h host] [-p
            prompt] [-u user] file ...
```

**Comment by jimi-c** [comment]
That's odd, if you change the IP to a different machine does it return the same error?

**Comment by wadey** [comment]
It looks like that last comment was a red herring, I think the `print " ".join` wasn't accurately representing how it was being executed.

I messed around running the EXEC command the same way that python would run it, and found that if I double escape the `"` chars, then it works:

```
>>> p = subprocess.Popen(['ssh', '-C', '-tt', '-q', 'trusty', '/bin/sh -c \'sudo -k && sudo -H -S -p "[sudo via ansible, key=bcabrlytlqgyhpodcrxoympxmchuenmh] password: " -u root /bin/sh -c \'"\'"\'echo SUDO-SUCCESS-bcabrlytlqgyhpodcrxoympxmchuenmh; /usr/bin/python /tmp/ansible-tmp-1398393337.43-167875872547355/command\'"\'"\'\''])
>>> bash: -c: line 0: unexpected EOF while looking for matching `"'
bash: -c: line 1: syntax error: unexpected end of file

>>> p = subprocess.Popen(['ssh', '-C', '-tt', '-q', 'trusty', '/bin/sh -c \'sudo -k && sudo -H -S -p \\"[sudo via ansible, key=bcabrlytlqgyhpodcrxoympxmchuenmh] password: \\" -u root /bin/sh -c \'\\"\'\\"\'echo SUDO-SUCCESS-bcabrlytlqgyhpodcrxoympxmchuenmh; /usr/bin/python /tmp/ansible-tmp-1398393337.43-167875872547355/command\'\\"\'\\"\'\''])
>>> SUDO-SUCCESS-bcabrlytlqgyhpodcrxoympxmchuenmh
{"changed": true, "end": "2014-04-24 19:54:22.531127", "stdout": "hello", "cmd": ["echo", "hello"], "start": "2014-04-24 19:54:22.529527", "delta": "0:00:00.001600", "stderr": "", "rc": 0}
```

The only difference is that the second time I changed all of the `"` to `\\"`. Is there some issue with the escaping?

**Comment by adamreid** [comment]
I was unable to reproduce from OS X 10.9.2 on a remote Ubuntu 14.04 box (also tried a CentOS box). I tried with both 1.5.5 and master. It looks like your environment is to blame.

**Comment by wadey** [comment]
Ok, I figured it out. I found if I ran the same command that failed from a different box, it worked.

I then found out that my machine had a `/usr/local/bin/ssh` wrapper installed that was to blame (it was prematurely evaluating the quotes).

Sorry for the confusing bug report, I wonder if there is anything ansible could have done to point me quicker at the error but I'm not sure that would be possible.

Thanks for the help @jimi-c and @adamreid.
